bug: promql.fragile disregards `for` zero value

This alert [intentionally](https://github.com/kubernetes-monitoring/kubernetes-mixin/pull/1014#discussion_r1947085111) has a `for` value of `0s` (due to 15m rate):

```yaml
  - "alert": "KubeNodeEviction"
    "annotations":
      "description": "Node {{ $labels.node }} is evicting Pods due to {{ $labels.eviction_signal }}.  Eviction occurs when eviction thresholds are crossed, typically caused by Pods exceeding RAM/ephemeral-storage limits."
      "runbook_url": "https://github.com/kubernetes-monitoring/kubernetes-mixin/tree/master/runbook.md#alert-name-kubenodeeviction"
      "summary": "Node is evicting pods."
    "expr": |
      sum(rate(kubelet_evictions{job="kubelet"}[15m])) by(cluster, eviction_signal, instance)
      * on (cluster, instance) group_left(node)
      max by (cluster, instance, node) (
        kubelet_node_name{job="kubelet"}
      )
      > 0
    "for": "0s"
    "labels":
      "severity": "info"
```

However it causes the following pint failure:

```
 Warning: fragile query (promql/fragile)
  ---> prometheus_alerts.yaml:686-691 -> `KubeNodeEviction`
686 |       sum(rate(kubelet_evictions{job="kubelet"}[15m])) by(cluster, eviction_signal, instance)
687 |       * on (cluster, instance) group_left(node)
688 |       max by (cluster, instance, node) (
689 |         kubelet_node_name{job="kubelet"}
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ This query can cause false positives when Prometheus restarts, add `for` option to avoid that.
690 |       )
691 |       > 0
make: *** [Makefile:80: pint-lint] Error 1
Error: Process completed with exit code 2.
```

IMO Pint should see a `for` value set explicitly to `0s` as intentionally, rather than missing.

--- a/docs/changelog.md
+++ b/docs/changelog.md
@@ -1,5 +1,11 @@
 # Changelog
 
+## v0.74.3
+
+### Fixed
+
+- Fixed false positive reports from [promql/fragile](checks/promql/fragile.md) check - [#1466](https://github.com/cloudflare/pint/issues/1466).
+
 ## v0.74.2
 
 ### Fixed

--- a/internal/checks/promql_fragile.go
+++ b/internal/checks/promql_fragile.go
@@ -117,7 +117,12 @@ func (c FragileCheck) checkPartialData(expr parser.PromQLExpr, src utils.Source,
 	}
 
 	for _, j := range src.Joins {
+		// Only look for joins that are aggregations.
 		if j.Src.Type != utils.AggregateSource {
+			continue
+		}
+		// Ignore joins that are not conditional and instead are used to add labels.
+		if len(j.AddedLabels) > 0 && !j.Src.IsConditional {
 			continue
 		}
 		if j.Depth > 0 {

--- a/internal/checks/promql_fragile_test.go
+++ b/internal/checks/promql_fragile_test.go
@@ -189,6 +189,22 @@ func TestFragileCheck(t *testing.T) {
 			checker:    newFragileCheck,
 			prometheus: noProm,
 		},
+		{
+			description: "false positive max - issues 1466",
+			content: `
+- alert: KubeNodeEviction
+  expr: |
+    sum(rate(kubelet_evictions{job="kubelet"}[15m])) by(cluster, eviction_signal, instance)
+    * on (cluster, instance) group_left(node)
+    max by (cluster, instance, node) (
+      kubelet_node_name{job="kubelet"}
+    )
+    > 0
+  for: 0s
+`,
+			checker:    newFragileCheck,
+			prometheus: noProm,
+		},
 	}
 
 	runTests(t, testCases)

--- a/internal/parser/utils/source.go
+++ b/internal/parser/utils/source.go
@@ -118,11 +118,12 @@ func newSource() Source {
 }
 
 type Join struct {
-	On       []string
-	Ignoring []string
-	Src      Source              // The source we're joining with.
-	Op       promParser.ItemType // The binary operation used for this join.
-	Depth    int                 // Zero if this is a direct join, non-zero otherwise. sum(foo * bar) would be in-direct join.
+	On          []string
+	Ignoring    []string
+	AddedLabels []string
+	Src         Source              // The source we're joining with.
+	Op          promParser.ItemType // The binary operation used for this join.
+	Depth       int                 // Zero if this is a direct join, non-zero otherwise. sum(foo * bar) would be in-direct join.
 }
 
 type Unless struct {
@@ -923,11 +924,12 @@ func parseBinOps(expr string, n *promParser.BinaryExpr) (src []Source) {
 					}
 				}
 				ls.Joins = append(ls.Joins, Join{
-					Src:      rs,
-					Op:       n.Op,
-					Depth:    0,
-					On:       onLabels(n.VectorMatching),
-					Ignoring: ignoringLabels(n.VectorMatching),
+					Src:         rs,
+					Op:          n.Op,
+					Depth:       0,
+					On:          onLabels(n.VectorMatching),
+					Ignoring:    ignoringLabels(n.VectorMatching),
+					AddedLabels: n.VectorMatching.Include,
 				})
 			}
 			ls.IsConditional, ls.ReturnInfo.IsReturnBool = checkConditions(ls, n.Op, n.ReturnBool)
@@ -968,11 +970,12 @@ func parseBinOps(expr string, n *promParser.BinaryExpr) (src []Source) {
 					}
 				}
 				rs.Joins = append(rs.Joins, Join{
-					Src:      ls,
-					Op:       n.Op,
-					Depth:    0,
-					On:       onLabels(n.VectorMatching),
-					Ignoring: ignoringLabels(n.VectorMatching),
+					Src:         ls,
+					Op:          n.Op,
+					Depth:       0,
+					On:          onLabels(n.VectorMatching),
+					Ignoring:    ignoringLabels(n.VectorMatching),
+					AddedLabels: n.VectorMatching.Include,
 				})
 			}
 			rs.IsConditional, rs.ReturnInfo.IsReturnBool = checkConditions(rs, n.Op, n.ReturnBool)
@@ -1010,11 +1013,12 @@ func parseBinOps(expr string, n *promParser.BinaryExpr) (src []Source) {
 					}
 				}
 				ls.Joins = append(ls.Joins, Join{
-					Src:      rs,
-					Op:       n.Op,
-					Depth:    0,
-					On:       onLabels(n.VectorMatching),
-					Ignoring: ignoringLabels(n.VectorMatching),
+					Src:         rs,
+					Op:          n.Op,
+					Depth:       0,
+					On:          onLabels(n.VectorMatching),
+					Ignoring:    ignoringLabels(n.VectorMatching),
+					AddedLabels: n.VectorMatching.Include,
 				})
 			}
 			ls.IsConditional, ls.ReturnInfo.IsReturnBool = checkConditions(ls, n.Op, n.ReturnBool)
@@ -1068,11 +1072,12 @@ func parseBinOps(expr string, n *promParser.BinaryExpr) (src []Source) {
 					})
 				case n.Op != promParser.LOR:
 					ls.Joins = append(ls.Joins, Join{
-						Src:      rs,
-						Op:       n.Op,
-						Depth:    0,
-						On:       onLabels(n.VectorMatching),
-						Ignoring: ignoringLabels(n.VectorMatching),
+						Src:         rs,
+						Op:          n.Op,
+						Depth:       0,
+						On:          onLabels(n.VectorMatching),
+						Ignoring:    ignoringLabels(n.VectorMatching),
+						AddedLabels: n.VectorMatching.Include,
 					})
 				}
 			}

